# Imports
## Panda e Numpy


Panda e numpy são bibliotecas auxiliares para cálculo e plot gráfico
Nós as usaremos como pd e np em diante


In [67]:
import pandas as pd
import numpy as np

## Seaborn e SKLearn

Seaborn pode nos ajudar com a visualização do dataset para posterior treinamento com sklearn.
SKLearn será o principal módulo do trabalho, fazendo a importação do tipo de aprendizado, treinamento do mesmo, e previsão com base no dataset


In [68]:
import seaborn as sns
import sklearn as sk

## Data Set 

O data set deve ser importado nessa fase.


In [69]:
dataset = pd.read_csv('db/movie_dataset.csv', sep=',',  index_col=0, na_values='?')

### Banco de dados e Objetivo
O "Movies Dataset" é uma extensa coleção de informações sobre 4.803 filmes, abrangendo uma ampla gama de detalhes, como orçamento, gêneros, empresas de produção, data de lançamento, receita, duração, idioma, popularidade, entre outros. Este conjunto de dados é composto por colunas distintas que representam cada um desses elementos, onde cada filme é uma instância única identificada por um ID. Além dos detalhes mencionados, também são fornecidas informações sobre as palavras-chave associadas ao filme, o resumo do enredo, o elenco, a equipe de produção, o diretor e o status do filme. As informações são de diversos tipos de dados, incluindo inteiros, objetos e números de ponto flutuante. O objetivo deste trabalho é utilizar este conjunto de dados para prever o custo de produção de um filme com base em diferentes combinações desses parâmetros, ajudando a entender como as características da produção podem impactar o orçamento necessário para produzir um filme.

In [70]:
dataset

budget                                    genres  \
index                                                        
0      237000000  Action Adventure Fantasy Science Fiction   
1      300000000                  Adventure Fantasy Action   
2      245000000                    Action Adventure Crime   
3      250000000               Action Crime Drama Thriller   
4      260000000          Action Adventure Science Fiction   
...          ...                                       ...   
4798      220000                     Action Crime Thriller   
4799        9000                            Comedy Romance   
4800           0             Comedy Drama Romance TV Movie   
4801           0                                       NaN   
4802           0                               Documentary   

                                                homepage      id  \
index                                                              
0                            http://www.avatarmovie.com/   19995   
1           http://disney.go.com/disneypictures/pirates/     285   
2            http://www.sonypictures.com/movies/spectre/  206647   
3                     http://www.thedarkknightrises.com/   49026   
4                   http://movies.disney.com/john-carter   49529   
...                                                  ...     ...   
4798                                                 NaN    9367   
4799                                                 NaN   72766   
4800   http://www.hallmarkchannel.com/signedsealeddel...  231617   
4801                         http://shanghaicalling.com/  126186   
4802                                                 NaN   25975   

                                                keywords original_language  \
index                                                                        
0      culture clash future space war space colony so...                en   
1      ocean drug abuse exotic island east india trad...                en   
2             spy based on novel secret agent sequel mi6                en   
3      dc comics crime fighter terrorist secret ident...                en   
4      based on novel mars medallion space travel pri...                en   
...                                                  ...               ...   
4798   united states\u2013mexico barrier legs arms pa...                es   
4799                                                 NaN                en   
4800   date love at first sight narration investigati...                en   
4801                                                 NaN                en   
4802                obsession camcorder crush dream girl                en   

                                 original_title  \
index                                             
0                                        Avatar   
1      Pirates of the Caribbean: At World's End   
2                                       Spectre   
3                         The Dark Knight Rises   
4                                   John Carter   
...                                         ...   
4798                                El Mariachi   
4799                                  Newlyweds   
4800                  Signed, Sealed, Delivered   
4801                           Shanghai Calling   
4802                          My Date with Drew   

                                                overview  popularity  \
index                                                                  
0      In the 22nd century, a paraplegic Marine is di...  150.437577   
1      Captain Barbossa, long believed to be dead, ha...  139.082615   
2      A cryptic message from Bond’s past sends him o...  107.376788   
3      Following the death of District Attorney Harve...  112.312950   
4      John Carter is a war-weary, former military ca...   43.926995   
...                                                  ...         ...   
4798   El Mariachi just wants to play his guitar and ...   14.2

In [71]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4803 entries, 0 to 4802
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4775 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4391 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status                4803

### Tratamento dos dados
Para trabalhar com esse dataset devemos mudar alguns dados e excluir outros. O nome do filme, overview e a homepage serão removidos. O cast e diretores, serão identificados por números assim como o diretor. Só serão aceitos filmes já lançados também. Outro ponto de seleção é o fato de que somente nos importa os filmes os quais já foram lançados por isso vamos fazer o filtro dos não lançados ainda

#### Get dummies 
Precisamos trabalhar mais o dataset antes de utilizar o get dummies, caso contrário o dataset acabará apresentando mais de 25 mil colunas para classificação

In [72]:
# Remoção das colunas não utilizáveis
dataset = dataset.loc[:, ~dataset.columns.isin(['original_title', 'overview', 'tagline', 'title', 'homepage', 'crew', 'keywords', 'original_language', 'id'])]
dataset = dataset.loc[dataset['status'] == 'Released']
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4795 entries, 0 to 4802
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4795 non-null   int64  
 1   genres                4767 non-null   object 
 2   popularity            4795 non-null   float64
 3   production_companies  4795 non-null   object 
 4   production_countries  4795 non-null   object 
 5   release_date          4794 non-null   object 
 6   revenue               4795 non-null   int64  
 7   runtime               4793 non-null   float64
 8   spoken_languages      4795 non-null   object 
 9   status                4795 non-null   object 
 10  vote_average          4795 non-null   float64
 11  vote_count            4795 non-null   int64  
 12  cast                  4754 non-null   object 
 13  director              4767 non-null   object 
dtypes: float64(3), int64(3), object(8)
memory usage: 561.9+ KB


In [73]:
dataset_base = dataset.loc[:, dataset.columns.isin(['budget', 'revenue', 'popularity', 'runtime', 'vote_average', 'vote_count'])]
dataset_base



budget  popularity     revenue  runtime  vote_average  vote_count
index                                                                      
0      237000000  150.437577  2787965087    162.0           7.2       11800
1      300000000  139.082615   961000000    169.0           6.9        4500
2      245000000  107.376788   880674609    148.0           6.3        4466
3      250000000  112.312950  1084939099    165.0           7.6        9106
4      260000000   43.926995   284139100    132.0           6.1        2124
...          ...         ...         ...      ...           ...         ...
4798      220000   14.269792     2040920     81.0           6.6         238
4799        9000    0.642552           0     85.0           5.9           5
4800           0    1.444476           0    120.0           7.0           6
4801           0    0.857008           0     98.0           5.7           7
4802           0    1.929883           0     90.0           6.3          16

[4795 rows x 6 columns]

In [74]:
dataset_dummies = dataset
dataset_dummies = pd.get_dummies(dataset_dummies)
dataset_dummies

budget  popularity     revenue  runtime  vote_average  vote_count  \
index                                                                         
0      237000000  150.437577  2787965087    162.0           7.2       11800   
1      300000000  139.082615   961000000    169.0           6.9        4500   
2      245000000  107.376788   880674609    148.0           6.3        4466   
3      250000000  112.312950  1084939099    165.0           7.6        9106   
4      260000000   43.926995   284139100    132.0           6.1        2124   
...          ...         ...         ...      ...           ...         ...   
4798      220000   14.269792     2040920     81.0           6.6         238   
4799        9000    0.642552           0     85.0           5.9           5   
4800           0    1.444476           0    120.0           7.0           6   
4801           0    0.857008           0     98.0           5.7           7   
4802           0    1.929883           0     90.0           6.3          16   

       genres_Action  genres_Action Adventure  \
index                                           
0              False                    False   
1              False                    False   
2              False                    False   
3              False                    False   
4              False                    False   
...              ...                      ...   
4798           False                    False   
4799           False                    False   
4800           False                    False   
4801           False                    False   
4802           False                    False   

       genres_Action Adventure Animation Comedy Family  \
index                                                    
0                                                False   
1                                                False   
2                                                False   
3                                                False   
4                                                False   
...                                                ...   
4798                                             False   
4799                                             False   
4800                                             False   
4801                                             False   
4802                                             False   

       genres_Action Adventure Animation Comedy Science Fiction  ...  \
index                                                            ...   
0                                                  False         ...   
1                                                  False         ...   
2                                                  False         ...   
3                                                  False         ...   
4                                                  False         ...   
...                                                  ...         ...   
4798                                               False         ...   
4799                                               False         ...   
4800                                               False         ...   
4801                                               False         ...   
4802                                               False         ...   

       director_Zackary Adler  director_Zak Penn  director_Zal Batmanglij  \
index                                                                       
0                       False              False                    False   
1                       False              False                    False   
2                       False              False                    False   
3                       False              False                    False   
4                       False              False                    False   
...                       ...                ...                      ...   
4798                    False       

# Uso do conjunto de dados

Queremos dar a opção ao usuário de citar alguns nomes que podem influenciar muito no budget no filme, como atores e diretores famosos e a empresa que está produzindo

## Usando o dataset sem personalidades famosas

Primeiro treinaremos a IA em um conjunto de dados o qual não possui os gêneros, atores e diretores para obter resultados que só dependam dos fatores presentes no dataset_base

Sendo assim podemos então selecionar os dados que queremos prever

In [75]:
noise = np.random.randint(0, 101, size=dataset_base['revenue'].shape[0])
dataset_base['revenue'] = dataset_base['revenue'] + noise

rev_pred_paramX = dataset_base.loc[:, dataset_base.columns != 'revenue'] 
rev_pred_paramY = np.array(dataset_base.loc[:, dataset_base.columns == 'revenue']).ravel()

/tmp/ipykernel_22022/1704839766.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset_base['revenue'] = dataset_base['revenue'] + noise


## Preparação do dataset

Para utilizar os algoritmos de predição vamos dividir a renda em 4 categorias, very low, low, medium, high

In [76]:
rev_pred_paramY = pd.Series(rev_pred_paramY)

In [77]:
rev_pred_paramY['revenue'] = pd.qcut(rev_pred_paramY, q=4, labels=['very low', 'low', 'medium', 'high'])

le = sk.preprocessing.LabelEncoder()
rev_pred_paramY = le.fit_transform(rev_pred_paramY['revenue'])
rev_pred_paramY

array([0, 0, 0, ..., 1, 3, 1])

## Treino e teste para àrvore de decisão

Fazemos então a separação desse conjunto de dados em treino e teste

In [78]:
rev_pred_paramX.fillna(1, inplace=True)
rev_pred_paramY = np.nan_to_num(rev_pred_paramY, nan=1)

X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(rev_pred_paramX, rev_pred_paramY, test_size=0.20, stratify=rev_pred_paramY, random_state=42)


/tmp/ipykernel_22022/1317501224.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rev_pred_paramX.fillna(1, inplace=True)


Importamos o método de árvore de decisão

In [79]:
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier(criterion='gini', random_state=42)

clf.fit(X_train, y_train)
predict = clf.predict(X_test)

E verificamos a acurácia

In [80]:
from sklearn.metrics import accuracy_score

acc = accuracy_score(y_test, predict)
print(f'A acurácia do Classificador é: {acc :.2%}')

A acurácia do Classificador é: 50.78%


Podemos observar que a acurácia do classificador DecisionTreeClassifier para o conjunto de dados escolhido é superior a 50% no conjunto de dados com 4 opções de escolha


## Treino e teste para GridSearch

In [81]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

parameters = {'n_neighbors' : [1, 3, 5, 7, 9, 11, 13], 'weights' : ['uniform', 'distance']}

knn = KNeighborsClassifier()

gs = GridSearchCV(knn, parameters, cv=5, scoring='accuracy')
gs.fit(rev_pred_paramX, rev_pred_paramY)

results = pd.DataFrame(gs.cv_results_)
view = ['params', 'mean_test_score', 'std_test_score', 'rank_test_score']
results[view].sort_values(by='rank_test_score')

params  mean_test_score  \
0     {'n_neighbors': 1, 'weights': 'uniform'}         0.329301   
1    {'n_neighbors': 1, 'weights': 'distance'}         0.329301   
2     {'n_neighbors': 3, 'weights': 'uniform'}         0.328676   
3    {'n_neighbors': 3, 'weights': 'distance'}         0.319499   
5    {'n_neighbors': 5, 'weights': 'distance'}         0.314077   
7    {'n_neighbors': 7, 'weights': 'distance'}         0.312617   
4     {'n_neighbors': 5, 'weights': 'uniform'}         0.308446   
9    {'n_neighbors': 9, 'weights': 'distance'}         0.304692   
6     {'n_neighbors': 7, 'weights': 'uniform'}         0.302190   
11  {'n_neighbors': 11, 'weights': 'distance'}         0.299896   
8     {'n_neighbors': 9, 'weights': 'uniform'}         0.298436   
13  {'n_neighbors': 13, 'weights': 'distance'}         0.294473   
10   {'n_neighbors': 11, 'weights': 'uniform'}         0.292388   
12   {'n_neighbors': 13, 'weights': 'uniform'}         0.288217   

    std_test_score  rank_test_score  
0         0.036185                1  
1         0.036185                1  
2         0.029950                3  
3         0.037351                4  
5         0.034631                5  
7         0.036106                6  
4         0.040047                7  
9         0.040468                8  
6         0.045878                9  
11        0.041167               10  
8         0.042420               11  
13        0.041968               12  
10        0.038846               13  
12        0.046980               14

Podemos observar então que a acurácia é extremamente baixa com este método também

## Multi-layer perceptron network

In [82]:
from sklearn.preprocessing import MinMaxScaler
# reescala os valores entre 0 e 1 utilizando o valor minimo e máximo de acada atributo
rev_pred_paramX_scaled = pd.DataFrame(MinMaxScaler().fit_transform(rev_pred_paramX), columns=rev_pred_paramX.columns)
rev_pred_paramX_scaled

budget  popularity   runtime  vote_average  vote_count
0     0.623684    0.171815  0.479290          0.72    0.858057
1     0.789474    0.158846  0.500000          0.69    0.327225
2     0.644737    0.122635  0.437870          0.63    0.324753
3     0.657895    0.128272  0.488166          0.76    0.662158
4     0.684211    0.050169  0.390533          0.61    0.154450
...        ...         ...       ...           ...         ...
4790  0.000579    0.016298  0.239645          0.66    0.017307
4791  0.000024    0.000734  0.251479          0.59    0.000364
4792  0.000000    0.001650  0.355030          0.70    0.000436
4793  0.000000    0.000979  0.289941          0.57    0.000509
4794  0.000000    0.002204  0.266272          0.63    0.001163

[4795 rows x 5 columns]

In [83]:
from sklearn.neural_network import MLPClassifier

parameters = {'hidden_layer_sizes' : [(5), (8), (15), (5, 3), (8, 5), (10, 5)],
              'max_iter' : [3000], 'random_state' : [42]}

mlp = MLPClassifier()
gs_mlp = GridSearchCV(mlp, parameters, cv=5, scoring='accuracy')

gs_mlp.fit(rev_pred_paramX_scaled, rev_pred_paramY)

view = ['params', 'mean_test_score', 'std_test_score', 'rank_test_score']
results = pd.DataFrame(gs_mlp.cv_results_)
results[view].sort_values(by='rank_test_score')

params  mean_test_score  \
0  {'hidden_layer_sizes': 5, 'max_iter': 3000, 'r...         0.516580   
1  {'hidden_layer_sizes': 8, 'max_iter': 3000, 'r...         0.515120   
3  {'hidden_layer_sizes': (5, 3), 'max_iter': 300...         0.509281   
2  {'hidden_layer_sizes': 15, 'max_iter': 3000, '...         0.507821   
5  {'hidden_layer_sizes': (10, 5), 'max_iter': 30...         0.506361   
4  {'hidden_layer_sizes': (8, 5), 'max_iter': 300...         0.500313   

   std_test_score  rank_test_score  
0        0.111110                1  
1        0.124864                2  
3        0.109489                3  
2        0.120596                4  
5        0.108608                5  
4        0.093939                6

# Inicio dos testes com os atores e diretores

## Performando alguns ajustes ao dataset

In [90]:
noise = np.random.randint(0, 101, size=dataset_dummies['revenue'].shape[0])
dataset_dummies['revenue'] = dataset_dummies['revenue'] + noise

dataset_dummies_rev_class = pd.qcut(dataset_dummies['revenue'], q=4, labels=['very low', 'low', 'medium', 'high'])
dataset_dummies['revenue'] = dataset_dummies_rev_class

TypeError: Object with dtype category cannot perform the numpy op add

In [85]:
rev_pred_paramX = dataset_dummies.loc[:, dataset_dummies.columns != 'revenue'] 
rev_pred_paramY = np.array(dataset_dummies.loc[:, dataset_dummies.columns == 'revenue']).ravel()
rev_pred_paramX.fillna(1, inplace=True)
rev_pred_paramY = np.nan_to_num(rev_pred_paramY, nan=1)

/tmp/ipykernel_22022/1703271108.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rev_pred_paramX.fillna(1, inplace=True)


## Árvore de decisão

In [88]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(rev_pred_paramX, rev_pred_paramY, test_size=0.20, stratify=rev_pred_paramY, random_state=42)


clf = DecisionTreeClassifier(criterion='gini', random_state=42)

clf.fit(X_train, y_train)
predict = clf.predict(X_test)

acc = accuracy_score(y_test, predict)
print(f'A acurácia do Classificador é: {acc :.2%}')

A acurácia do Classificador é: 59.12%


## Grid Search com Kneighbors

In [87]:
parameters = {'n_neighbors' : [1, 3, 5, 7, 9, 11, 13], 'weights' : ['uniform', 'distance']}

knn = KNeighborsClassifier()

gs = GridSearchCV(knn, parameters, cv=5, scoring='accuracy')
gs.fit(rev_pred_paramX, rev_pred_paramY)

results = pd.DataFrame(gs.cv_results_)
view = ['params', 'mean_test_score', 'std_test_score', 'rank_test_score']
results[view].sort_values(by='rank_test_score')

params  mean_test_score  \
0     {'n_neighbors': 1, 'weights': 'uniform'}         0.335766   
1    {'n_neighbors': 1, 'weights': 'distance'}         0.335766   
2     {'n_neighbors': 3, 'weights': 'uniform'}         0.326799   
3    {'n_neighbors': 3, 'weights': 'distance'}         0.320542   
5    {'n_neighbors': 5, 'weights': 'distance'}         0.316580   
7    {'n_neighbors': 7, 'weights': 'distance'}         0.312409   
4     {'n_neighbors': 5, 'weights': 'uniform'}         0.308863   
9    {'n_neighbors': 9, 'weights': 'distance'}         0.304275   
6     {'n_neighbors': 7, 'weights': 'uniform'}         0.303858   
11  {'n_neighbors': 11, 'weights': 'distance'}         0.300938   
8     {'n_neighbors': 9, 'weights': 'uniform'}         0.300104   
13  {'n_neighbors': 13, 'weights': 'distance'}         0.297810   
10   {'n_neighbors': 11, 'weights': 'uniform'}         0.292179   
12   {'n_neighbors': 13, 'weights': 'uniform'}         0.288634   

    std_test_score  rank_test_score  
0         0.039871                1  
1         0.039871                1  
2         0.029799                3  
3         0.038146                4  
5         0.029179                5  
7         0.031522                6  
4         0.034385                7  
9         0.036233                8  
6         0.040647                9  
11        0.038095               10  
8         0.041156               11  
13        0.039923               12  
10        0.039665               13  
12        0.043459               14

## Multi-layer perceptron network
Podemos observar que houve grande aumento de tempo de execução para o teste como número de colunas tão alto e o multi-layer perceptron network é bem intesivo no consumo então ele não roda em um tempo relativamente aceitável

# Fim dos testes e baixa acurácia

Os testes com as formas de machine learning de árvore de decisão e multi-layer perceptron apresentaram uma acurácia < 54%. 

O método com os K neighbours e o conjunto de dados com os atores diferentes obteve 33% de mean_test_score o que é um desempenho muito inferior

Devemos então analisar e trabalhar com os dados para melhorar a nossa acurácia